include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,271 @@
+      time_by_index_t time_by_index_;
+
+      double resample_time_ = 0.0;
+
+      tuner_t(resource_with_index_t br, size_type resources_size, double rt)
+        : t0_(std::chrono::steady_clock::now()),
+          best_resource_(br),
+          max_resource_to_profile_(resources_size),
+          resample_time_(rt) {}
+
+      size_type get_resource_to_profile() {
+        std::lock_guard<std::mutex> l(m_);
+        if (next_resource_to_profile_ < 2*max_resource_to_profile_) {
+          // do everything twice
+          return next_resource_to_profile_++ % max_resource_to_profile_;
+        } else if (resample_time_ == never_resample) {
+          return use_best_resource;
+        } else {
+          auto now = std::chrono::steady_clock::now();
+          auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now-t0_).count();
+          if (ms < resample_time_) {
+            return use_best_resource;
+          } else {
+            t0_ = now;
+            return next_resource_to_profile_ = 0;
+          }
+        }
+      }
+
+      // called to add new profile info
+      void add_new_timing(resource_with_index_t r, timing_t t) {
+        std::unique_lock<std::mutex> l(m_);
+        auto index = r.index_;
+        timing_t new_value = t;
+        if (time_by_index_.count(index) == 0) {
+          // ignore the 1st timing to cover for JIT compilation
+          time_by_index_[index] = time_data_t{0, std::numeric_limits<timing_t>::max()};
+        } else {
+          auto &td = time_by_index_[index];
+          auto n = td.num_timings_;
+          new_value = (n*td.value_+t)/(n+1);
+          td.num_timings_ = n+1;
+          td.value_ = new_value;
+        }
+        if (new_value < best_timing_) {
+          best_timing_ = new_value;
+          best_resource_ = r;
+        }
+      }
+
+    };
+
+    class auto_tune_selection_type {
+      using policy_t = auto_tune_policy<Backend, KeyArgs...>;
+      policy_t policy_;
+      resource_with_index_t resource_;
+      std::shared_ptr<tuner_t> tuner_;
+
+    public:
+      auto_tune_selection_type() : policy_(deferred_initialization) {}
+
+      auto_tune_selection_type(const policy_t& p, resource_with_index_t r, std::shared_ptr<tuner_t> t)
+        : policy_(p), resource_(r), tuner_(::std::move(t)) {}
+
+      auto unwrap() { return ::oneapi::dpl::experimental::unwrap(resource_.r_); }
+
+      policy_t get_policy() { return policy_; };
+
+      void report(const execution_info::task_time_t&, const typename execution_info::task_time_t::value_type& v) const {
+        tuner_->add_new_timing(resource_, v);
+      }
+    };
+
+   public:
+
+    // Needed by Policy Traits
+    using resource_type = decltype(unwrap(std::declval<wrapped_resource_t>()));
+    using wait_type = typename Backend::wait_type;
+    using selection_type = auto_tune_selection_type;
+
+    auto_tune_policy(deferred_initialization_t) {}
+
+    auto_tune_policy(double resample_time=never_resample) {
```

Comment:
Agreed to change `double` to `uint64_t` because currently the API is not user friendly. For example:" `0.1` resample_time time looks like a "killing parameter" for the policy instance.